sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpClient.cs

```diff
@@ -155,7 +153,7 @@ internal class AmqpClient : TransportClient
                 }.Uri;
 
                 EventHubName = eventHubName;
-                Credential = credential;
+                Credential = new EventHubTokenCredential(credential, ServiceEndpoint.ToString());
```

Comment:
> This would also necessitate some minor changes to `AmqpConnectionScope` so that it also takes only an `EventHubTokenCredential`

I am going to be bluntly honest with you, I did blindly copy that solution 😄

Reply:
> We could unify that by just changing the base for [SharedAccessSignatureCredential](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/Authorization/SharedAccessSignatureCredential.cs), I think. No?

Do you mean, turning this hierarchy:
```
TokenCredential
     |-- SharedAccessSignatureCredential
     |-- EventHubSharedKeyCredential
     |-- EventHubTokenCredential
```

Into this hierarchy?

```
TokenCredential
     |--EventHubTokenCredential
         |-- SharedAccessSignatureCredential
         |-- EventHubSharedKeyCredential
```

The main challenge I can see would be populating the ```eventHubResource``` parameter in the constructor of ```EventHubTokenCredential``` when invoked from ```SharedAccessSignatureCredential``` or from ```EventHubSharedKeyCredential```?

```csharp
public EventHubTokenCredential(TokenCredential tokenCredential, string eventHubResource)
```

Reply:
Looking more closely, maybe we leave the hierarchy alone and just make sure that we wrap `SharedAccessSignatureCredential` as `EventHubTokenCredential`.   It looks like I had already accounted for that with the `IsSharedAccessSignatureCredential ` property,  we're just not consistently wrapping things in the `EventHubConnection`.    

I believe that if we modified two points in the connection, those should be the only paths needed since all client construction routes through there.  I'm looking at Lines [171](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubConnection.cs#L171) and [201](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubConnection.cs#L201).

 I **_think_** would be the only non-test changes, though I'll apologize for being a bit busy this morning and not being able to look as thoroughly as I would have liked.   Do you see anything that I'm overlooking that would make this high effort or would invalidate the approach?

Reply:
> I **_think_** would be the only non-test changes, though I'll apologize for being a bit busy this morning and not being able to look as thoroughly as I would have liked. Do you see anything that I'm overlooking that would make this high effort or would invalidate the approach?

No worries and wish you a nice morning 😄 

I think everything you said is correct.

The only thing that I might flag is that the code at line 171 would be a bit bulky?

```csharp
var sharedAccessSignature = new SharedAccessSignature
(
    BuildAudienceResource(connectionOptions.TransportType, fullyQualifiedNamespace, eventHubName),
    connectionStringProperties.SharedAccessKeyName,
    connectionStringProperties.SharedAccessKey
);

var sharedCredentials = new SharedAccessSignatureCredential(sharedAccessSignature);
var tokenCredentials = new EventHubTokenCredential(sharedCredentials, BuildAudienceResource(connectionOptions.TransportType, fullyQualifiedNamespace, eventHubName));

(..)

InnerClient = CreateTransportClient(fullyQualifiedNamespace, eventHubName, tokenCredentials, connectionOptions);
```

Also, changing the following signature will require a lot of refactoring test side and I will start looking at it:

```csharp
internal virtual TransportClient CreateTransportClient(string fullyQualifiedNamespace,
                                                       string eventHubName,
                                                       EventHubTokenCredential credential,
                                                       EventHubConnectionOptions options)
```

Reply:
Thank you very much for your guidance with this change.

I have pushed the changes.

I have started fixing the tests. Most fail as they mock a ```TokenCredential```.

Reply:
> Thank you very much for your guidance with this change.

I'd consider it more sharing my personal opinion, though I do have the advantage of knowing what I was thinking when I wrote the original stuff.   😛